yafut: add a kernel update tool for MikroTik NAND

kempniu opened the pull request:
## Background

Commit 9d96b6fb72 ("ath79/mikrotik: disable building NAND images") disabled building images for Mikrotik devices with NAND flash due to a less than satisfactory method used for updating the kernel on those devices back then.  The log message for the above commit contains a neat summary of the problems that this PR attempts to solve:

```
The current support for MikroTik NAND-based devices relies on a
gross hack that packs the kernel into a static YAFFS stub, as the
stock bootloader only supports booting a YAFFS-encapsulated kernel.

The problem with this approach is that since the kernel partition is
blindly overwritten without any kind of wear or badblock management
(due to lack of proper support for YAFFS in OpenWRT), the NAND flash
is not worn uniformly and eventually badblocks appear, leading to
unbootable devices.

This issue has been reported here [1] and discussed in more detail
here [2].

[1] https://forum.openwrt.org/t/rb433-bad-sector-cannot-start-openwrt/71519
[2] https://github.com/openwrt/openwrt/pull/3026#issuecomment-673597461

Until a proper fix is found (or the stock bootloader supports other
filesystems), we disable building these images to prevent unknowing
users from risking their devices.
```

## How this PR proposes to solve the problem

This PR adds support for updating the kernel on Mikrotik devices with NAND flash using a new tool, [Yafut](https://github.com/kempniu/yafut), which enables copying files from/to Yaffs file systems even if the kernel does not have native support for the Yaffs file system compiled in.  Instead of erasing the entire NAND partition holding the kernel during every system upgrade, Yafut preserves the Yaffs filesystem present on that partition and only replaces the kernel executable.  This allows bad block information to be preserved across sysupgrade runs and also enables wear leveling on the NAND partition holding the kernel.

The catch is that the tool relies on the MEMREAD ioctl for MTD character devices that was only first included in Linux 6.1.  That ioctl allows userspace applications to use the Linux kernel's OOB autoplacement mechanism while reading data from NAND devices.  This PR includes a 5.15 backport of the kernel patch series implementing the MEMREAD ioctl.

## Tests carried out so far

In the [tests I ran myself on a Mikrotik RB951Ui-2HnD router](https://github.com/openwrt/openwrt/pull/12226), Yafut was able to successfully replace both the stock RouterOS kernel when run from an OpenWRT initramfs and an OpenWRT kernel when run from an OpenWRT installation booted from NAND.

I do not own a device with bad blocks on the NAND flash, but I carried out some tests using nandsim that seem to show that Yafut is able to handle these just fine (of course as long as the number of available non-bad blocks on the MTD partition holding the kernel is large enough to fit the kernel file).

## How to test

The simplest way to test the tool on a given device is to try reading the kernel executable from the relevant MTD partition and then, if reading works, try to write the kernel executable back to the MTD, e.g.:

    cd /tmp; yafut -d /dev/mtd1 -r -i kernel -o kernel-copy && yafut -d /dev/mtd1 -w -i kernel-copy -o kernel

If the device boots fine after running the above commands and restarting, then updating the kernel via sysupgrade should work just as well.

If the above invocation is not able to read the `kernel` file from flash, try adding the `-T` command-line option to it.  This forces use of inband tags (see kempniu/yafut#10).  If adding this option enables the tool to read the `kernel` file, make sure to also add it to the invocation used for writing the `kernel` file back to the MTD.

## Other notes

  - ~~This PR does **not** switch any existing Mikrotik device to the new kernel update method as it has not been tested widely yet.  See #12226 for an [example](https://github.com/openwrt/openwrt/pull/12226/commits/3ebd5e91b1d22a684fb328b59856e4b02c806412#diff-1412db68ca9f904f7f6124a6786fecdce34b3afa21843b806241971e9198ac3e) of how to switch a specific device to Yafut.  (I do not own any currently-supported device that I could switch over to the new method immediately in this PR.)~~ This PR switches all MikroTik NAND devices to use Yafut because OpenWRT images for these devices are currently not built by default and this PR intends to change that. (See a discussion [below](#pullrequestreview-1359424572) for more information.)

  - NOR flash is currently not supported, but work is underway to change that.

john-tho commented:
Huge thanks for working on this. Well done. It might eventually allow us to simply boot OpenWrt from RouterBOOT on modern Mikrotik hardware as well (rb5009 & ipq60xx also use YAFFS).

To clarify (please correct where needed):
Tool requires an existing YAFFS filesystem on the device, and can write files to it.
Do you have a best practices method to getting to the point where we can use it? Example, reformat NAND in RouterBOOT, or netinstall RouterOS first, or just try it from any stage?

While not _needed_ for NOR, I will eventually try this on some of my NOR devices. In RouterBOOT v7 on NOR, the old method of booting `kernel` no longer works. The entire NPK is written to `bootimage`, which will be a >10MiB file in RouterOS. If I then boot an OpenWrt initramfs, where this YAFFS partition is cut short (in DTS), down to what I need for kernel, will `yafut` cope with this? Then, when I overwrite `bootimage` with a smaller file, will it clean up the no-longer-needed old blocks from the longer `bootimage`? May not even matter to RouterBOOT though.

Even further off-topic, I guess I could try this with RouterOS partitions: https://help.mikrotik.com/docs/display/ROS/Partitions, but will need to work out how they work, and modify my DTS to suit.

Cheers

john-tho commented on target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh:
```diff
@@ -27,6 +27,22 @@ platform_do_upgrade_mikrotik_nand() {
+	board_dir=${board_dir%/}
+	[ -n "$board_dir" ] || return
+
+	tar xf "$1" ${board_dir}/kernel -O | yafut -d "$fw_mtd" -w -i - -o kernel -m 0755 || echo "Kernel update failed with error $?"
```
Do we want to bail out, and avoid doing `nand_do_update` if there is an issue with the `yafut` write?

kempniu replied:
Right, that's a good point.  I simply mirrored what `platform_do_upgrade_mikrotik_nand()` (the `mtd erase` + `nandwrite` counterpart of this function) does and it did not seem to care a lot :-)  But yes, I think we should bail out if `yafut` reports an error.  Perhaps an even safer thing to do is to first read the kernel back from the MTD (to `/dev/null`) and only attempt writing if reading succeeds?  That should prevent messing up working devices if `yafut` gets confused for whatever reason.

kempniu replied:
> Perhaps an even safer thing to do is to first read the kernel back from the MTD (to `/dev/null`) and only attempt writing if reading succeeds?

Answering my own question: it probably is, but if we rely on this "sanity check", then `platform_do_upgrade_mikrotik_nand()` is not going to be able to write the kernel executable to an empty Yaffs filesystem, which is rather limiting.

For that reason, I simply changed the `echo ...` part above to a `return` in 67bf7e7908330139935bfdd84b6779428d7c28b0.

kempniu commented on target/linux/ath79/image/common-mikrotik.mk:
```diff
@@ -18,6 +18,6 @@ define Device/mikrotik_nand
   IMAGE/sysupgrade.bin = append-kernel | kernel2minor -s 2048 -e -c | \
 	sysupgrade-tar kernel=$$$$@ | append-metadata
-  DEVICE_PACKAGES := nand-utils
+  DEVICE_PACKAGES := nand-utils yafut
```
Ah, sure, I guess I just did not want the PR to be particularly intrusive, but your point makes sense to me, thanks.

kempniu replied:
Addressed as of 67bf7e7908330139935bfdd84b6779428d7c28b0, thanks.

kempniu commented on target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh:
```diff
@@ -7,7 +7,7 @@ platform_check_image() {
 }
 
-RAMFS_COPY_BIN='fw_printenv fw_setenv nandwrite'
+RAMFS_COPY_BIN='fw_printenv fw_setenv nandwrite yafut'
 RAMFS_COPY_DATA='/etc/fw_env.config /var/lock/fw_printenv.lock'
```
I think you're right.  a66eee63368eb5625b359244bfffc9b9eb921875 added these lines to `target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh` when it moved `platform_do_upgrade_mikrotik_nand()` from `target/linux/ath79/nand/base-files/lib/upgrade/platform.sh`, which also contained these lines.  However, they were added to the latter file by 55e6c903ae20bd8e694f4626166766e59602f87c, which added support for a GL.iNet device, not a Mikrotik one.  I will clean this up as you suggested.

kempniu replied:
Cleaned up in commit da6a964e81f6a246791ab160d8c49b733d46f059, thanks.

kempniu replied:
Oops, I have only just now noticed that the log message for da6a964e81f6a246791ab160d8c49b733d46f059 references a malformed commit hash (`66eee6336` should actually be a66eee6336).  I am going to unresolve this conversation so that I remember to fix this in one of the subsequent PR updates.

kempniu commented:
Thanks for taking a look, @john-tho.

> To clarify (please correct where needed): Tool requires an existing YAFFS filesystem on the device, and can write files to it. Do you have a best practices method to getting to the point where we can use it? Example, reformat NAND in RouterBOOT, or netinstall RouterOS first, or just try it from any stage?

I'd say just try it.  It should work both with a fully erased MTD partition and an existing Yaffs filesystem - after all, Yafut is not much more than a dumb bridge between original Yaffs code and MTD ioctls manipulating data on the NAND device, so whatever magic Yaffs code is available to pull off should be possible via Yafut as well.

> While not _needed_ for NOR, I will eventually try this on some of my NOR devices. In RouterBOOT v7 on NOR, the old method of booting `kernel` no longer works. The entire NPK is written to `bootimage`, which will be a >10MiB file in RouterOS. If I then boot an OpenWrt initramfs, where this YAFFS partition is cut short (in DTS), down to what I need for kernel, will `yafut` cope with this?

Sorry, I am not familiar with the `bootimage` thing and what "cutting the Yaffs partition short" means here.  Do you perhaps have any references (commits, docs) I could take a look at to better understand your question?

> Then, when I overwrite `bootimage` with a smaller file, will it clean up the no-longer-needed old blocks from the longer `bootimage`? May not even matter to RouterBOOT though.

Replacing any file causes Yaffs code to update the file size and therefore also mark any chunks "freed" this way as available for reuse.

f00b4r0 commented:
This is brilliant. It's an easy step toward getting rid of kernel2minor, which is good. It simplifies the process of building an image for Mikrotik NAND devices. The code is clean and abundantly commented, kudos.
I hope this PR gets merged soon.

robimarko commented:
If I got it right, `kernel2minor` is still required?

Also, `yafut: add a kernel update tool for Mikrotik NAND` should probably be split to include only the tool and then recipe changes should be separate.

f00b4r0 commented:
> If I got it right, `kernel2minor` is still required?

Not that I can tell. See first reply in this comment: https://github.com/openwrt/openwrt/pull/12225#issuecomment-1477322421

The new device recipe in #12226 does not use kernel2minor.

> Also, `yafut: add a kernel update tool for Mikrotik NAND` should probably be split to include only the tool and then recipe changes should be separate.

Agreed

john-tho commented on package/utils/yafut/Makefile:
```diff
@@ -0,0 +1,35 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=yafut
+PKG_RELEASE:=$(AUTORELEASE)
```
I _think_ that OpenWrt was looking to move away from `AUTORELEASE`. There is a check in the packages branch against it: https://github.com/openwrt/packages/blob/master/.github/workflows/check-autorelease-deprecation.yml
```suggestion
PKG_RELEASE:=1
```

kempniu replied:
Ah, of course, thanks, I wasn't aware of this. Will fix.

kempniu replied:
Addressed as of ca52c328051a02888e18354be2d58f1ff96ff738, thanks.

kempniu commented:
@f00b4r0, thank you for taking a look and for your kind words.  I hope this works out.

@robimarko, thank you for taking a look as well.  As @f00b4r0 suggested [above](#issuecomment-1484156667), the approach proposed in this PR does not rely on `kernel2minor` in any way and intends to be a replacement for the latter in the long term.  However, it seemed to me that removing `kernel2minor` immediately (in this PR) would be a bit premature, given the new tool's lack of exposure to a broader set of Mikrotik devices, so I decided to tread carefully for the time being.

If the lack of reliance on `kernel2minor` is not immediately obvious from the commit message, perhaps that is a sign that it should be tweaked to make things more clear?

> > Also, `yafut: add a kernel update tool for Mikrotik NAND` should probably be split to include only the tool and then recipe changes should be separate.
> 
> Agreed

Do I understand correctly that you are suggesting to separate the addition of `package/utils/yafut/Makefile` from the updates to `target/linux/ath79/mikrotik/` into separate *commits*?  Or a separate *pull request*?

f00b4r0 commented:
> @f00b4r0, thank you for taking a look and for your kind words. I hope this works out.
> 
> @robimarko, thank you for taking a look as well. As @f00b4r0 suggested [above](#issuecomment-1484156667), the approach proposed in this PR does not rely on `kernel2minor` in any way and intends to be a replacement for the latter in the long term. However, it seemed to me that removing `kernel2minor` immediately (in this PR) would be a bit premature, given the new tool's lack of exposure to a broader set of Mikrotik devices, so I decided to tread carefully for the time being.

kernel2minor will still be needed for NOR devices for the time being. The approach proposed here would not be applicable (these use a unified `firmware` mtd partition, and splitting that would actually be a less flexible regression.

If yafut can be massaged to output a yaffs stub to a file as kernel2minor does, it would then be able to replace kernel2minor entirely and that would be a huge win. Hint hint ;)

> If the lack of reliance on `kernel2minor` is not immediately obvious from the commit message, perhaps that is a sign that it should be tweaked to make things more clear?

It may be a good idea to emphasise this yes.

> > > Also, `yafut: add a kernel update tool for Mikrotik NAND` should probably be split to include only the tool and then recipe changes should be separate.
> > 
> > 
> > Agreed
> 
> Do I understand correctly that you are suggesting to separate the addition of `package/utils/yafut/Makefile` from the updates to `target/linux/ath79/mikrotik/` into separate _commits_? Or a separate _pull request_?

Separate commits will do.

Cheers

kempniu commented:
> kernel2minor will still be needed for NOR devices for the time being. The approach proposed here would not be applicable (these use a unified `firmware` mtd partition, and splitting that would actually be a less flexible regression).
> 
> If yafut can be massaged to output a yaffs stub to a file as kernel2minor does, it would then be able to replace kernel2minor entirely and that would be a huge win. Hint hint ;) (we don't need to update the yaffs in situ on NOR devices since bad blocks are essentially a "practically never happen" thing and we can live with this).

To be honest, I am rather fond of the "do one thing, and do it well" philosophy.  Extending a file copy utility with file system image generation capabilities is a bit at odds with that approach :-)  I was not aware that getting rid of `kernel2minor` is a goal per se - from what I gathered, the problem to solve was that of bad block handling and wear leveling on NAND devices. Perhaps the [`mkyaffs2image` utility in the Yaffs repository](http://www.aleph1.co.uk/gitweb/?p=yaffs2.git;a=blob;f=utils/mkyaffs2image.c;h=92f3f61b5e370178ce8a84e05c096be32ac213e4;hb=HEAD) could be tweaked to do what `kernel2minor` does?

> > If the lack of reliance on `kernel2minor` is not immediately obvious from the commit message, perhaps that is a sign that it should be tweaked to make things more clear?
> 
> It may be a good idea to emphasise this yes.

Sure, will do.

> > > > Also, `yafut: add a kernel update tool for Mikrotik NAND` should probably be split to include only the tool and then recipe changes should be separate.
> > > 
> > > 
> > > Agreed
> > 
> > 
> > Do I understand correctly that you are suggesting to separate the addition of `package/utils/yafut/Makefile` from the updates to `target/linux/ath79/mikrotik/` into separate _commits_? Or a separate _pull request_?
> 
> Separate commits will do.

Sure, I will take care of that.

f00b4r0 commented:
> > kernel2minor will still be needed for NOR devices for the time being. The approach proposed here would not be applicable (these use a unified `firmware` mtd partition, and splitting that would actually be a less flexible regression).
> > If yafut can be massaged to output a yaffs stub to a file as kernel2minor does, it would then be able to replace kernel2minor entirely and that would be a huge win. Hint hint ;) (we don't need to update the yaffs in situ on NOR devices since bad blocks are essentially a "practically never happen" thing and we can live with this).
> 
> To be honest, I am rather fond of the "do one thing, and do it well" philosophy.

That's a good philosophy which I share, so I can't fault you for that :)

> Extending a file copy utility with file system image generation capabilities is a bit at odds with that approach :-)

Agreed, I was merely throwing an idea around ;-)

> I was not aware that getting rid of `kernel2minor` is a goal per se

If you look at that code you'll probably figure out why it would be nice to get there.

> - from what I gathered, the problem to solve was that of bad block handling and wear leveling on NAND devices.

Correct, that's the problem your PR solves and it solves it very nicely. Again I was just pushing my own wishes in the mix, don't think they're relevant here: they're not :-)

Thanks again for your excellent contribution.

robimarko commented on target/linux/ath79/image/common-mikrotik.mk:
```diff
@@ -21,3 +21,8 @@ define Device/mikrotik_nand
   DEFAULT := n
 endef
+
+define Device/mikrotik_nand_yafut
```
Shouldn't `Device/mikrotik_nand` profile be edited to get rid of the custom recipe that uses kernel2minor instead of adding new recipe that is not used?

robimarko replied:
To me its not intrusive as currently no NAND images are generated, and since this tool finally allows them to be used again recipe should be updated.

@f00b4r0 Thoughts?

f00b4r0 replied:
> To me its not intrusive as **currently no NAND images are generated**, and since this tool finally allows them to be used again recipe should be updated.

That's a good point which I failed to consider: we no longer generate any NAND image, so there's no "switching over" per se. It would certainly speed up testing too; so I think we can do this. @kempniu sorry for the trouble :)

f00b4r0 replied:
@kempniu ath79 "supported" Mikrotik NAND devices are fairly consistent. I would think it's "okay" to switch them all over. Worst case scenario is a softbrick, these are "easy" to debrick (the only redeeming feature of RouterBoot I guess). So I'd suggest applying the change to all ath79 mikrotik NAND devices and see what happens. We're not touching release yet anyway, this can see some broader testing in master this way.

@robimarko makes a strong point that this would allow these devices to be properly supported again (if yafut does work reliably, which master testing will show).

kempniu replied:
Fair enough.  All I'm trying to say is that (at this point) Yafut should be considered alpha-quality software that is capable of setting hardware on fire, causing PCB components to spontaneously emit toxic gases, tearing holes in spacetime and whatnot.  Your point about debricking being relatively easy makes sense to me, though, so does the part about master vs. release. I'll adjust the PR accordingly.

xback replied:
> Yafut should be considered alpha-quality software that is capable of setting hardware on fire, causing PCB components to spontaneously emit toxic gases, tearing holes in spacetime and whatnot.

Thanks for the heads-up.
I'll let a junior engineer test this initially then over here.


Jokes aside,
I've got a ton of these boards (from very old to the latest revisions, different nand chips on some of them), so I'm happy to do extensive testing if it aids the cause.

kempniu replied:
4a0bbce8471ebd0c15af43aa9f6ebd89c7f8d674 modifies the `mikrotik_nand` template instead of adding a separate one.  It also restores building MikroTik NAND images by defaults.  The commit message was updated accordingly.

f00b4r0 commented:
@xback you might be interested in this btw :)

xback commented:
> @xback you might be interested in this btw :)

For sure!
I'm following this with great interest.

kempniu commented:
@f00b4r0, following up on our earlier discussion [above](#issuecomment-1484206835), I [looked around a bit](https://github.com/kempniu/yafut/issues/4#issuecomment-1487043587) and I believe Yafut could be used with NOR flash without adding file system image generation capabilities to it.  Instead, after implementing some NOR-specific tweaks, it could be used this way:

    mtd erase <partition-name>
    yafut -d /dev/mtdX -w -i <path-to-bootimage> -o bootimage

Would this be acceptable?  AFAICT, this is practically equivalent to generating an image of a Yaffs file system containing a single file and flashing that to the MTD, which is what happens for NOR devices now, isn't it?  Or am I confused?  (I admittedly lack a broader perspective on the boot process used by various MikroTik devices these days, so my apologies if the above is rubbish.)

Note the "tricky part" mentioned in the Yafut issue comment I linked to above does not apply for an erased NOR flash because pretty much any sane set of parameters can be used for a fresh file system.

f00b4r0 commented:
> @f00b4r0, following up on our earlier discussion [above](#issuecomment-1484206835), I [looked around a bit](https://github.com/kempniu/yafut/issues/4#issuecomment-1487043587) 

Thanks! But don't bother yourself too much with this at this stage, let's focus on the NAND side of things first :)

> and I believe Yafut could be used with NOR flash without adding file system image generation capabilities to it. Instead, after implementing some NOR-specific tweaks, it could be used this way:
> 
> ```
> mtd erase <partition-name>
> yafut -d /dev/mtdX -w -i <path-to-bootimage> -o bootimage
> ```
> 
> Would this be acceptable?

Unfortunately I think not. Consider that on NOR devices, the DTS defines a global `firmware` partition to which the kernel + rootfs is written by sysupgrade. In other words, the kernel partition is sized to exactly fit the image kernel (modulo padding as needed), nothing more nothing less.

The rootfs is squashfs, not yaffs, so if I understand correctly what you suggest we would have to write kernel and rootfs separately, handling alignment/boundaries, which is significantly less trivial than the current method.

> AFAICT, this is practically equivalent to generating an image of a Yaffs file system containing a single file and flashing that to the MTD, which is what happens for NOR devices now, isn't it? Or am I confused? (I admittedly lack a broader perspective on the boot process used by various MikroTik devices these days, so my apologies if the above is rubbish.)
> 
> Note the "tricky part" mentioned in the Yafut issue comment I linked to above does not apply for an erased NOR flash because pretty much any sane set of parameters can be used for a fresh file system.

Indeed, none of this is completely trivial and as I said, it's not a high priority for now, we can always revisit this topic later :)

Thanks!

davidskongri commented:
The first version of this commit left my Mikrotik SXT Lite 5 in a bootloop state, plus it left my equipment with the LEDs literally blown. The smell of burning was felt throughout the room. The strange thing is that the equipment continued to work perfectly after this, and I currently have it running OpenWrt 19.0.7, so I still have doubts if it was after flashing the equipment or it was because it was already in bad shape.

  The other question I have is about how to flash. For example, currently my Mikrotik SXT Lite 5 is running with OpenWrt 19.0.7, to flash your own build from the master branch with this pull request you would have to do it directly from OpenWrt 19.0.7 or you would have to load the OpenWrt image on the computer memory and then flash?

kempniu commented:
> > Would this be acceptable?
> 
> Unfortunately I think not. Consider that on NOR devices, the DTS defines a global `firmware` partition to which the kernel + rootfs is written by sysupgrade _in one go_. In other words, the kernel partition is sized to exactly fit the image kernel (modulo padding as needed), nothing more nothing less.

Okay, right, thanks for keeping me honest, I got confused by the [Yafut issue](https://github.com/kempniu/yafut/issues/4) that was related to NOR devices and I have only now dug a bit deeper into how the whole OpenWRT sysupgrade machinery works on NOR devices.  I think I understand earlier comments on this PR better now :-)

I also agree that it is not as simple as I initially thought and that my naïve `mtd erase` + `yafut` proposal above does not cut it.

> The rootfs is squashfs, not yaffs, so if I understand correctly what you suggest we would have to write kernel and rootfs separately, handling alignment/boundaries, which is significantly less trivial than the current method.

Right, I still have a few ideas for working around these obstacles, but I agree that we should not hijack this PR with NOR discussions and focus on getting the most burning problem (NAND) solved first.

kempniu commented:
> The first version of this commit left my Mikrotik SXT Lite 5 in a bootloop state, plus it left my equipment with the LEDs literally blown. The smell of burning was felt throughout the room.

I knew it!!!

> The strange thing is that the equipment continued to work perfectly after this, and I currently have it running OpenWrt 19.0.7, so I still have doubts if it was after flashing the equipment or it was because it was already in bad shape.

Oh...

> The other question I have is about how to flash. For example, currently my Mikrotik SXT Lite 5 is running with OpenWrt 19.0.7, to flash your own build from the master branch with this pull request you would have to do it directly from OpenWrt 19.0.7 or you would have to load the OpenWrt image on the device memory via TFTP server using the Reset Button and then flash?

The latter (via initramfs).  However, note I still need to update this PR so that it modifies the existing `mikrotik_nand` template rather than creating a separate one (`mikrotik_nand_yafut`) that one would need to manually switch devices over to.

That's only needed for the sysupgrade part, though.  If you feel like doing some quick experiments in the meantime, feel free to just load the initramfs built from the current state of this PR (better update `PKG_SOURCE_VERSION` in `package/utils/yafut/Makefile` to `72586787b99a6e87ddb23e2aebfcfcdc8c097ff4` first, though - important bug fix) and try reading the kernel file using `yafut` and writing it back to the flash (see OP).  If your device still works after this (i.e. boots OpenWRT 19.07, your currently-flashed version), Yafut should also work with sysupgrade when this PR is ready.  But be prepared to reflash your device if the tool messes up.

If that doesn't work, you can also give Yafut's `force-inband-tags` branch a shot; it's currently at `454113da8447f8a659c27b20e3b2cf96d34da101`.  That branch might be gone soon, though (when I'm done with implementing autodetection of inband tags).

davidskongri commented:
@kempniu Okay. I will compile a new image and try. But with the new commits the device revived the LEDs jjjj, but still in a bootloop state. The resulting initramfs image doesn't boot either, I think it's mostly because it turned out to be too big (about 18MB).

  Now I will make a light build to test. In the morning when I try I pass the results.

kempniu committed:
yafut: add a kernel update tool for MikroTik NAND

Commit 9d96b6fb72 ("ath79/mikrotik: disable building NAND images")
disabled building images for MikroTik devices with NAND flash due to a
less than satisfactory method used for updating the kernel on those
devices back then.

To address the problem, add support for updating the kernel on MikroTik
devices with NAND flash using a new tool, Yafut, which enables copying
files from/to Yaffs file systems even if the kernel does not have native
support for the Yaffs file system compiled in.  Instead of erasing the
entire NAND partition holding the kernel during every system upgrade
(which is what the previously-used approach employing kernel2minor
involved), Yafut preserves the Yaffs filesystem present on that
partition and only replaces the kernel executable.  This allows bad
block information to be preserved across sysupgrade runs and also
enables wear leveling on the NAND partition holding the kernel.  Yafut
does not rely on kernel2minor in any way and intends to eventually
supersede the latter for NAND devices.

Signed-off-by: Michał Kępień <openwrt@kempniu.pl>

kempniu committed:
ath79: mikrotik: drop unused files from ramdisk

The ramdisk used by sysupgrade on MikroTik devices currently includes
U-Boot fw_* files that are not necessary for performing a system upgrade
on that platform.  The relevant lines were added to
target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh by commit
a66eee6336 ("ath79: add mikrotik subtarget"), likely because they also
existed in target/linux/ath79/nand/base-files/lib/upgrade/platform.sh,
where the platform_do_upgrade_mikrotik_nand() function moved by commit
a66eee6336 originally lived.  However, these lines were added to
target/linux/ath79/nand/base-files/lib/upgrade/platform.sh by commit
55e6c903ae ("ath79: GL-AR300M: provide NAND support; increase to 4 MB
kernel"), which is not related to MikroTik devices in any way.

Remove the code adding unused U-Boot fw_* files to the ramdisk used by
sysupgrade on MikroTik devices.

Signed-off-by: Michał Kępień <openwrt@kempniu.pl>

kempniu committed:
kernel: backport MEMREAD ioctl

MEMREAD is a new ioctl for MTD character devices that was first included
in Linux 6.1.  It allows userspace applications to use the Linux
kernel's OOB autoplacement mechanism while reading data from NAND
devices.  The Yafut tool needs this ioctl to do its job.

Signed-off-by: Michał Kępień <openwrt@kempniu.pl>

kempniu commented:
As of 4a0bbce8471ebd0c15af43aa9f6ebd89c7f8d674, this PR includes the most recent Yafut revision available (kempniu/yafut@16435e89d449f953712983315e1a89cdb678620d).  It contains several important bug fixes compared to the original version and adds support for forcing inband tags (many thanks to @john-tho for testing), which may be necessary on some devices.  I updated the "How to test" section in OP to mention that.

I am currently unaware of any outstanding bugs for NAND devices, so it feels like the right moment to do some more testing "in the wild".  (I will start working on NOR flash support in the meantime, but it is not a priority for now.)

robimarko commented:
Great, it would be nice to see testing in the wild

f00b4r0 commented:
@xback care to merge this and #12226? ;)

rogerpueyo commented:
Hi, @kempniu,

Thanks for this PR! I could only quickly try your code on a RB922 I had lying around. First I compiled flashed latest master and flashed it. Then I compiled your PR and sysupgraded it twice. On a first check, everything seemed to be fine. I will give it another shot in a few days, though.

BR

hauke commented on target/linux/ath79/mikrotik/target.mk:
```diff
@@ -3,7 +3,7 @@ FEATURES += minor nand
 IMAGES_DIR := ../../..
 
-DEFAULT_PACKAGES += wpad-basic-mbedtls
+DEFAULT_PACKAGES += wpad-basic-mbedtls yafut
```
Why do you add `yafut` to all mikrotik devices and not only the nand devices like it iwas done with `nand-utils` before?

hauke replied:
I just read you added it here to have it in the ramdisk too.
Please add this to the commit description.

kempniu replied:
Sure, thanks.  Does e322e5268948f8ff88b2246536a5e360fa0236b2 work for you?  (I applied the `mikrotik_nor` tweak you suggested and updated the second paragraph of the log message compared to commit 4a0bbce8471ebd0c15af43aa9f6ebd89c7f8d674.)

hauke replied:
looks good.

hauke commented on target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh:
```diff
@@ -20,8 +20,7 @@ platform_do_upgrade_mikrotik_nand() {
 
-	mtd erase kernel
-	tar xf "$1" ${board_dir}/kernel -O | nandwrite -o "$fw_mtd" -
+	tar xf "$1" ${board_dir}/kernel -O | yafut -d "$fw_mtd" -w -i - -o kernel -m 0755 || return
```
Could this fail?
What happens when the Yaffs is partly broken?

kempniu replied:
> Could this fail?

It can, if Yaffs code reaches the conclusion that it is unable to write the `kernel` file to the MTD, for whatever reason.  In particular, this can happen if the MTD partition holding the kernel is using different Yaffs parameters than the defaults assumed by Yafut.  This is why earlier revisions of this PR (e.g. dacf09232647d2742fedc2f59f96d93e7e7732c2) added a separate helper function to `target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh` that devices could be switched to once someone verifies that the new kernel update method works fine for a given device.  The feedback I [heard](#discussion_r1149585644) was that MikroTik NAND images are currently broken anyway, so it's okay to switch them to a new tool that may help solve the problem and see what happens.

If the Yaffs parameters used by Yafut are the same as the parameters used by the existing file system on the MTD partition holding the kernel, writing should not fail (assuming of course that the kernel file fits on the given MTD partition).

> What happens when the Yaffs is partly broken?

As it is often the case, it depends.  The most reasonable answer I can give is that what happens in such a case would be similar to mounting the relevant MTD partition using a kernel that has Yaffs support compiled in and running `mv /path/to/kernel /mnt/yaffs-partition/kernel`.  If the file system is "not that broken", Yaffs code may be able to fix it.  Yafut does not contain any extra Yaffs-related logic that the upstream Yaffs code would not already contain.  If things are utterly broken, `mtd erase kernel` and creating a fresh file system may be a feasible last-resort option.

> Is yafut handling a partly damaged file system?

I think my answer to the previous question also applies here.

In short: once the proper set of Yaffs parameters that the bootloader on a given device expects is determined and Yafut is told to use them, I would expect things to be stable indefinitely.

If I misunderstood your questions, please let me know and I'll be happy to explain further.

john-tho replied:
>  What happens when the Yaffs is partly broken?

If the YAFFS is modified by `yafut` on NAND, but the write is incomplete (possible example, kernel too big: not sure if we test this in build, or should test before sysupgrade), device will not boot OpenWrt. Recoverable through BOOTP with button access. If the kernel ELF header is broken, device will normally attempt BOOTP automatically.

---

Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`. Example usage here: https://github.com/openwrt/openwrt/blob/787ecdf66de6798540fb8b756404566341b4733c/target/linux/ipq40xx/base-files/lib/upgrade/platform.sh#L7
`yafut` read should work unless kernel on NAND is broken (example, NAND erased), and then device should not be able to boot NAND anyway, unless `yafut` is having an error reading (example, wrong geometry). From my quick look through https://github.com/openwrt/openwrt/blob/787ecdf66de6798540fb8b756404566341b4733c/package/base-files/files/usr/libexec/validate_firmware_image#L59 it seems sysupgrade would still be forceable, so we could then notify user.

It would make the upgrade take longer for the read, and we are not expecting ath79 NAND devices to vary NAND geometry. I think wait for more feedback before implementing this?

Cheers

f00b4r0 replied:
> Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`. Example usage here:

I think that's quite overkill. Remember that up to that point, we were blindly writing a raw blob (containing a yaffs fs) to the underlying device without any kind of check. Surely a tool that actually handles the filesystem is a major step up already.

> `yafut` read should work unless kernel on NAND is broken (example, NAND erased), and then device should not be able to boot NAND anyway, unless `yafut` is having an error reading (example, wrong geometry). 

Remember that all currently "supported" ath79 NAND devices (128+MB NAND) use the same geometry, as evidenced by the same set of kernel2minor parameters being used for all these devices (and further evidenced by having a single image fitting all devices in the ar71xx era).

My 2c: let's not make this more complicated than it has to be ;)

kempniu replied:
> Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`.

@john-tho Obviously this is not the worst idea - we even briefly discussed it [above](#pullrequestreview-1349537607) - but I am leaning towards the conclusion that this would not be a good safety measure.  In the experiments I run during Yafut development, it happened sometimes (when the Yaffs parameters were messed up or when the code had bugs that were ironed out later on) that the tool seemed to have read a file from the MTD correctly - exit code 0, no obvious errors reported, correct file size - but the *contents* of the file were malformed.  Given that there is no way to know what the *expected* contents of the `kernel` file are, adding more complexity for reading it before attempting to overwrite it does not seem to be worth the benefits.

> Remember that all currently "supported" ath79 NAND devices (128+MB NAND) use the same geometry, as evidenced by the same set of kernel2minor parameters being used for all these devices (and further evidenced by having a single image fitting all devices in the ar71xx era).

@f00b4r0 This is a good point.  The only thing I can think of that could go wrong is that the MTD partition holding the kernel could theoretically use a different ECC layout than the one selected by the Linux kernel for the NAND chip.  Given that I only own two (very similar) devices, it is impossible for me to assess whether that is something that happens out there.

f00b4r0 replied:
> @f00b4r0 This is a good point. The only thing I can think of that could go wrong is that the MTD partition holding the kernel could theoretically use a different ECC layout than the one selected by the Linux kernel for the NAND chip. 

I don't see how: this would result in failed mtd writes regardless of underlying filesystem. The fact that a single image was suitable to work with all ar71xx NAND devices suggests there is no such problem. Again, I think we shouldn't speculate too much, or this PR will never get merged.

In my opinion this PR is **ready for merge now**: the resulting system is *better in every way* to the prior alternative (images not built and potential brick due to lack of wear-management).

kempniu replied:
I could continue the ECC discussion, but supporting your conclusion presented above feels more productive :-)  Perfect is indeed the enemy of good.

kempniu committed:
ath79: mikrotik: update kernel on NAND using Yafut

Instead of erasing the entire NAND partition holding the kernel during
every system upgrade and then flashing a Yaffs file system image
prepared using kernel2minor (not accounting for bad blocks in the
process), use the Yafut utility to replace the kernel executable on
MikroTik NAND devices, preserving the existing Yaffs file system
(including bad block information) on the partition holding the kernel.

Add Yafut to DEFAULT_PACKAGES for the ath79/mikrotik target, so that the
tool is included in the initramfs images created when building for
multiple profiles.  However, exclude Yafut from the images built for
MikroTik devices with NOR flash as the tool is currently only meant to
be used on devices with NAND flash.

As this addresses the concerns for MikroTik NAND devices discussed in
commit 9d96b6fb72 ("ath79/mikrotik: disable building NAND images"),
re-enable building images for these devices.

Signed-off-by: Michał Kępień <openwrt@kempniu.pl>

Borromini commented:
@robimarko @f00b4r0 Can we get this into master or is something missing in its present state?

Would be great if this can get merged before the 23.xx fork, or is this too risky?

f00b4r0 commented:
@Borromini ideally this would be merged asap so it can see some further exposure in master "long enough" before 23.xx is branched.

robimarko commented:
@Borromini Neither of us has commit rights, if it was up to me I would have already merged it.

robimarko commented:
@Ansuel Well, that is something that wasn't being checked before as well, I would say that it's not a blocker for now.

Ansuel commented:
was curious if there were some dry-run feature in the tool since in theory all the bad block info are known prior flashing

f00b4r0 commented:
@Ansuel maybe this can be added in a later revision of the tool. In the meantime, this PR provides major improvements over the current situation with _zero_ regression: can it be merged already, so that it spends enough time in master before the upcoming branch?

It would be good to have this in 23.x, thus reenabling support for several boards that have been disabled back in 21.02.

xback commented:
Purely informational..
I've done a test upgrading a board which contained a nand image from previous method, and tried sysupgrade using this new method:

Looks like it dies during flashing.
The device is unbootable afterwards.


```
Watchdog handover: fd=3
- watchdog -
Watchdog did not previously reset the system
Fri Apr 21 09:09:13 UTC 2023 upgrade: Sending TERM to remaining processes ...
Fri Apr 21 09:09:13 UTC 2023 upgrade: Sending signal TERM to Cm2_Main (2054)
Fri Apr 21 09:09:13 UTC 2023 upgrade: Sending signal TERM to dnsmasq (2286)
Fri Apr 21 09:09:17 UTC 2023 upgrade: Sending KILL to remaining processes ...
Fri Apr 21 09:09:17 UTC 2023 upgrade: Sending signal KILL to Cm2_Main (2054)
[ 5722.666819] stage2 (3104): drop_caches: 3
Fri Apr 21 09:09:23 UTC 2023 upgrade: Switching to ramdisk...
[ 5726.820110] UBIFS (ubi0:2): background thread "ubifs_bgt0_2" stops
[ 5726.836992] UBIFS (ubi0:2): un-mount UBI device 0
Fri Apr 21 09:09:27 UTC 2023 upgrade: Performing system upgrade...
Unlocking kernel ...
Erasing kernel ...
Writing data to block 0 at offset 0x0
Writing data to block 1 at offset 0x20000
Writing data to block 2 at offset 0x40000
Writing data to block 3 at offset 0x60000
Writing data to block 4 at offset 0x80000
Writing data to block 5 at offset 0xa0000
Writing data to block 6 at offset 0xc0000
Writing data to block 7 at offset 0xe0000
Writing data to block 8 at offset 0x100000
Writing data to block 9 at offset 0x120000
Writing data to block 10 at offset 0x140000
Writing data to block 11 at offset 0x160000
Writing data to block 12 at offset 0x180000
Writing data to block 13 at offset 0x1a0000
Writing data to block 14 at offset 0x1c0000
Writing data to block 15 at offset 0x1e0000
Writing data to block 16 at offset 0x200000
Writing data to block 17 at offset 0x220000
Writing data to block 18 at offset 0x240000
Writing data to block 19 at offset 0x260000
Unexpected EOF. Expecting at least 32 more bytes for OOB
nandwrite: error!: Data was only partially written due to error
           error 0 (No error information)
verifying sysupgrade tar file integrity
umount: can't unmount /dev/ubiblock0_1: Invalid argument
[ 5729.657293] block ubiblock0_1: released
Volume ID 0, size 95 LEBs (12062720 bytes, 11.5 MiB), LEB size 126976 bytes (124.0 KiB), dynamic, name "rootfs", alignment 1
Set volume size to 110850048
Volume ID 1, size 873 LEBs (110850048 bytes, 105.7 MiB), LEB size 126976 bytes (124.0 KiB), dynamic, name "rootfs_data", alignment 1
[ 5734.172668] UBIFS (ubi0:1): default file-system created
[ 5734.179263] UBIFS (ubi0:1): Mounting in unauthenticated mode
[ 5734.185217] UBIFS (ubi0:1): background thread "ubifs_bgt0_1" started, PID 3931
[ 5734.314551] UBIFS (ubi0:1): UBIFS: mounted UBI device 0, volume 1, name "rootfs_data"
[ 5734.322694] UBIFS (ubi0:1): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[ 5734.332792] UBIFS (ubi0:1): FS size: 109453312 bytes (104 MiB, 862 LEBs), max 873 LEBs, journal size 5459968 bytes (5 MiB, 43 LEBs)
[ 5734.344799] UBIFS (ubi0:1): reserved for root: 4952683 bytes (4836 KiB)
[ 5734.351513] UBIFS (ubi0:1): media format: w5/r0 (latest is w5/r0), UUID 763609DA-11D8-4348-B6FA-0B69BAAA30C0, small LPT model
[ 5734.385372] UBIFS (ubi0:1): un-mount UBI device 0
[ 5734.390203] UBIFS (ubi0:1): background thread "ubifs_bgt0_1" stops
configuration saved
sysupgrade successful
umount: can't unmount /dev: Resource busy
umount: can't unmount /tmp: Resource busy
[ 5734.420711] reboot: Restarting system

```

f00b4r0 commented:
> Purely informational.. I've done a test upgrading a board which contained a nand image from previous method, and tried sysupgrade using this new method:

Are you sure your test was with the new method? See:

[...]
> Writing data to block 16 at offset 0x200000
> Writing data to block 17 at offset 0x220000
> Writing data to block 18 at offset 0x240000
> Writing data to block 19 at offset 0x260000
> Unexpected EOF. Expecting at least 32 more bytes for OOB
> **nandwrite**: error!: Data was only partially written due to error
>            error 0 (No error information)

The new method does not use `nandwrite`.

xback commented:
My bad, I could have elaborated a bit more:

Starting condition:
- Board running a nand image generated before the yafut series.
- create new image using yafut series
- try flashing the new generated image using sysupgrade (which still uses nandwrite at this point)

robimarko commented:
Yeah, that is not going to work, you need an image with yafut and updated sysupgrade recipe

f00b4r0 commented:
Ah I see. AIUI that's not supported, see https://github.com/openwrt/openwrt/pull/12225#issuecomment-1487944134

xback commented:
So .. In my case where I have hundreds of rb922's in the field, it will be impossible to upgrade now ;-)

f00b4r0 commented:
You can upgrade; you just have to do it from initramfs.

kempniu commented:
The reason it does not work is because this PR [changes](https://github.com/openwrt/openwrt/pull/12225/files#diff-a7d66f368074047c5a42b56b479c9427adc9cc048d1df696fc528c0e017ab33d) what the `kernel` file inside the sysupgrade tar file contains:

  - before this PR, `kernel` was a static Yaffs stub,
  - after this PR, `kernel` is the kernel executable.

I'm sure you can see how trying to write an executable to an MTD partition using `nandwrite` is not a great idea ;)

As for hundreds of devices in the field, I guess you could prepare a custom sysupgrade image with the hunk linked to above reverted that would be a "stepping stone": you could then first sysupgrade to such an image (which would contain the `yafut` executable and a kernel version with all the necessary ioctls backported), using the only sysupgrade method that the previously-deployed device knows about (`nandwrite`).  Reboot and only then upgrade "properly" using a `yafut`-based image from `master`.  I have not tested this, though, and I certainly may have missed some obstacle that would prevent this working.  Just thinking out loud.

xback commented:
@kempniu
@robimarko 
@f00b4r0 
Thanks all for taking the time to think along :-)

xback commented:
I can confirm using an intermediate image in between works.
